File: services/node-service/src/routes/health.js
const express = require('express');
const router = express.Router();
const config = require('../config');
const { checkPythonHealth } = require('../services/pythonClient');
const { getActiveClientCount } = require('../services/websocketServer');

const startTime = Date.now();

/**
 * Detailed Aggregated Health Check for Kubernetes and CI/CD Verification
 */
router.get('/health', async (req, res) => {
  const uptimeSeconds = Math.round((Date.now() - startTime) / 1000);
  const memory = process.memoryUsage();
  
  // Check downstream Python service
  const pythonStatus = await checkPythonHealth();

  const isHealthy = pythonStatus.status === 'UP';
  const overallStatus = isHealthy ? 'healthy' : 'degraded';

  const responsePayload = {
    status: overallStatus,
    service: config.serviceName,
    version: config.version,
    timestamp: new Date().toISOString(),
    uptime_seconds: uptimeSeconds,
    system: {
      environment: config.environment,
      node_version: process.version,
      pid: process.pid,
      memory_rss_mb: Math.round((memory.rss / (1024 * 1024)) * 100) / 100,
      heap_used_mb: Math.round((memory.heapUsed / (1024 * 1024)) * 100) / 100,
      active_ws_clients: getActiveClientCount(),
    },
    dependencies: {
      python_analytics_service: pythonStatus,
    },
  };

  // Return 200 even if degraded so CI/CD can read json diagnostics, or 503 if strict unhealthy
  res.status(200).json(responsePayload);
});

/**
 * Kubernetes Liveness Probe:
 * Simple fast 200 response to confirm Node.js process is active
 */
router.get(['/health/live', '/healthz', '/livez'], (req, res) => {
  res.status(200).json({
    status: 'alive',
    service: config.serviceName,
    timestamp: new Date().toISOString(),
  });
});

/**
 * Kubernetes Readiness Probe:
 * Confirms gateway is ready to receive and route traffic
 */
router.get(['/health/ready', '/readyz'], (req, res) => {
  res.status(200).json({
    status: 'ready',
    service: config.serviceName,
    ready: true,
  });
});

module.exports = router;

File: services/python-service/app/routes/health.py
import time
import os
import psutil
from datetime import datetime, timezone
from fastapi import APIRouter, Response, status
from app.config import settings
from app.models.schemas import HealthResponse, HealthComponentStatus

router = APIRouter(tags=["Health"])

START_TIME = time.time()

@router.get("/health", response_model=HealthResponse)
async def get_health():
    """
    Detailed aggregated health endpoint for CI/CD and Kubernetes verification.
    """
    uptime = time.time() - START_TIME
    
    # System metrics
    try:
        process = psutil.Process(os.getpid())
        mem_info = process.memory_info()
        memory_mb = round(mem_info.rss / (1024 * 1024), 2)
        cpu_percent = process.cpu_percent(interval=None)
    except Exception:
        memory_mb = 0.0
        cpu_percent = 0.0

    return HealthResponse(
        status="healthy",
        service=settings.SERVICE_NAME,
        version=settings.VERSION,
        timestamp=datetime.now(timezone.utc).isoformat(),
        uptime_seconds=round(uptime, 2),
        system={
            "environment": settings.ENVIRONMENT,
            "memory_usage_mb": memory_mb,
            "cpu_percent": cpu_percent,
            "pid": os.getpid(),
        },
        components={
            "threat_engine": HealthComponentStatus(
                status="UP",
                details={"patterns_loaded": 4, "threshold": settings.ANOMALY_THRESHOLD}
            ),
            "telemetry_collector": HealthComponentStatus(
                status="UP",
                details={"status": "active"}
            )
        }
    )

@router.get("/health/live")
@router.get("/healthz")
@router.get("/livez")
async def liveness_probe():
    """
    Kubernetes Liveness Probe:
    Verifies if the service process is running and not deadlocked.
    """
    return {"status": "alive", "service": settings.SERVICE_NAME, "timestamp": datetime.now(timezone.utc).isoformat()}

@router.get("/health/ready")
@router.get("/readyz")
async def readiness_probe():
    """
    Kubernetes Readiness Probe:
    Verifies if the service is ready to accept incoming traffic.
    """
    return {"status": "ready", "service": settings.SERVICE_NAME, "ready": True}
